### ISSUE TITLE
How to train data on a different dataset

### ISSUE THREAD

**Issue by jksorger** [whjdark/AAGNet#1]
First of all hello and congrats on this really great work!
I was also testing hierarchical CADNet on their MFCAD++ dataset - this is how I accidentally came across your work too. I was training and testing successfully AAGNet on the dataset that you have provided (MFInstSeg) and I really like your results. Therefore I have tried to test a couple of my own CAD models and as far as I am creating them in FreeCAD it works great. However, when I use another (commercial) Software for STEP export for the same models, where for whichever reason almost everything is being converted to BSplines (and therefore STEP files look much different than with OpenCascade/FreeCAD) the results are not as expected anymore. I assume I would need to train a model again on such a dataset, but I am missing quite some information what I would need to do. I would very grateful if you would provide full information and would be also very motivated to contribute to AAGNet if I could.   
Thanks a lot in advance and best regards, Jovana

**Comment by jksorger** [comment]
> First of all hello and congrats on this really great work! I was also testing hierarchical CADNet on their MFCAD++ dataset - this is how I accidentally came across your work too. I was training and testing successfully AAGNet on the dataset that you have provided (MFInstSeg) and I really like your results. Therefore I have tried to test a couple of my own CAD models and as far as I am creating them in FreeCAD it works great. However, when I use another (commercial) Software for STEP export for the same models, where for whichever reason almost everything is being converted to BSplines (and therefore STEP files look much different than with OpenCascade/FreeCAD) the results are not as expected anymore. I assume I would need to train a model again on such a dataset, but I am missing quite some information what I would need to do. I would very grateful if you would provide full information and would be also very motivated to contribute to AAGNet if I could. Thanks a lot in advance and best regards, Jovana

P.S. I have just realized that there is a documentation I have asked for, but the links are not working  ('\' instead of '/').

**Comment by whjdark** [comment]
> > First of all hello and congrats on this really great work! I was also testing hierarchical CADNet on their MFCAD++ dataset - this is how I accidentally came across your work too. I was training and testing successfully AAGNet on the dataset that you have provided (MFInstSeg) and I really like your results. Therefore I have tried to test a couple of my own CAD models and as far as I am creating them in FreeCAD it works great. However, when I use another (commercial) Software for STEP export for the same models, where for whichever reason almost everything is being converted to BSplines (and therefore STEP files look much different than with OpenCascade/FreeCAD) the results are not as expected anymore. I assume I would need to train a model again on such a dataset, but I am missing quite some information what I would need to do. I would very grateful if you would provide full information and would be also very motivated to contribute to AAGNet if I could. Thanks a lot in advance and best regards, Jovana
> 
> P.S. I have just realized that there is a documentation I have asked for, but the links are not working ('\' instead of '/').

Thank for your feedback, We have updated the README.md for fixing the wrong links.

If you want to prepare your own dataset, you can follow these:
1. prepare the training STEP files, and then use dataset/AAGExtractor.py to convert STEPs to gAAG representation.
2.  prepare your labels file, which is a big JSON containing all labels.
     The JSON stores a python list, which looks like: [ label1, label2, label3, ... , labeln].
      label1, label2, label3, ... , labeln are python lists, which contain two items: [ string, dict ]. 
      string is the name of your STEP (without the extension '.stp' or '.step')
      dict is a python dict, which contains 3 keys:
      The value of 'seg'  is a python dict containing segmentation labels, whose key is face id and value is the type id of feature (see dataset/dataset.png)
      The value of 'inst' is a 2D python list containing instance labels, which is face relation matrix _M_. For example, a slot feature has 3 faces and their ids are 14, 15, and 16, so there are links between the two of them. In the face relation matrix_M_, _M_[13, 13]. _M_[13, 14], _M_[13, 15], _M_[14, 14]. _M_[14, 15], _M_[15, 15] are set to 1.
     The value of 'bottom'  is a python dict containing bottom face labels, whose key is face id and value is 0-1 label. If the face is the bottom of a feature, the label should be 1.
3. prepare your data partition like MFInstseg_partition
4. organize your dataset like MFInstSeg
Besides, you can modify the dataloader for better supporting your dataset.

Please let us know if you have any further questions or suggestions.

**Comment by jksorger** [comment]
Thank you very much for your time to reply and all the infos. I would definitely like to test it for the following:
- MFCAD (or MFCAD++).
- Our own CAD Models
I have understood how the labeled data should look like in your case, but I am wondering how you generate it. I have not found this information in your paper (congrats once again, great work!). Are you initially using ASIN or some other solution to produce files with labels (and then correcting/improving it)? 
I must admit I am reluctant to do this work manually :). But if I would find some nice automatism for generating labels for my step files, perhaps I can provide you with some "non-synthetic" data, which you mention in you future work?
Please let me know if you have ideas/suggestions concerning that.

**Comment by whjdark** [comment]
> Thank you very much for your time to reply and all the infos. I would definitely like to retrain and test it for the following:
> 
>     * MFCAD (or MFCAD++).
> 
>     * Our own CAD Models
>       I have understood how the labeled data should look like in your case, but I am wondering how you generate it. I have not found this information in your paper (congrats once again, great work!). Are you initially using ASIN or some other solution to produce files with labels (and then correcting/improving it)?
>       I must admit I am reluctant to do this work manually :). But if I would find some nice automatism for generating labels for my step files, perhaps I can provide you with some "non-synthetic" data, which you mention in you future work?
>       Please let me know if you have ideas/suggestions concerning that.

1. We provide the method to test on MFCAD/MFCAD++ in our paper, and you are welcome to use it.

2. Our dataset generation method (including CAD models and labels) is a modification of the method from [the paper of hierarchical CADNET](https://gitlab.com/qub_femg/machine-learning/mfcad2-dataset), which uses the OpenCASCADE-based API.
First, we sketch a feature on a random face, then we extrude it along a certain direction, so we can know the newly created faces through the OpenCASCADE API. According to the definition of machining features, these new faces are the ones that belong to that feature, and we store their corresponding labels in JSON format as described in our paper.  (more details are shown in [shape_factory.py](https://github.com/whjdark/AAGNet/blob/main/dataset/Utils/shape_factory.py#L381), [machine_feature.py](https://github.com/whjdark/AAGNet/blob/main/dataset/Features/machining_features.py#L648), and [feature_creation.py](https://github.com/whjdark/AAGNet/blob/main/dataset/feature_creation.py#L262).)


3.We are very grateful if you can provide some non-synthetic data, which would be very helpful for our future work. In fact, manual data annotation is inevitable in learning-based methods, but there are some methods to reduce the labor, such as semi-automatic annotation, which we hope to provide in our future development. I hope this can help you.

**Comment by jksorger** [comment]
Thank you for your reply. Yes, it is clear one needs to do some of the annotation manually, but it is almost not feasible within "reasonable" time with big CAD-models, so it would be great if you could provide some tools to help with the annotation. Then we could also help with/ do the rest of the (manual) annotation for our (non-synthetic) models and would be glad to support your research this way.
